apps/gdalalg_vector_select.cpp

```diff
@@ -0,0 +1,335 @@
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+/************************************************************************/
+/*         GDALVectorSelectAlgorithm::GDALVectorSelectAlgorithm()       */
+/************************************************************************/
+
+GDALVectorSelectAlgorithm::GDALVectorSelectAlgorithm(bool standaloneStep)
+    : GDALVectorPipelineStepAlgorithm(NAME, DESCRIPTION, HELP_URL,
+                                      standaloneStep)
+{
+    AddArg("fields", 0, _("Fields to select (or exclude if --exclude)"),
+           &m_fields)
+        .SetPositional()
+        .SetRequired();
+    AddArg("exclude", 0, _("Exclude specified fields"), &m_exclude)
+        .SetMutualExclusionGroup("exclude-ignore");
+    AddArg("ignore-missing-fields", 0, _("Ignore missing fields"),
+           &m_ignoreMissingFields)
+        .SetMutualExclusionGroup("exclude-ignore");
+}
+
+namespace
+{
+
+/************************************************************************/
+/*                   GDALVectorSelectAlgorithmLayer                     */
+/************************************************************************/
+
+class GDALVectorSelectAlgorithmLayer final
+    : public OGRLayer,
+      public OGRGetNextFeatureThroughRaw<GDALVectorSelectAlgorithmLayer>
+{
+  private:
+    OGRLayer &m_oSrcLayer;
+    OGRFeatureDefn *const m_poFeatureDefn = nullptr;
+    std::vector<int> m_anMapSrcFieldsToDstFields{};
+    std::vector<int> m_anMapDstGeomFieldsToSrcGeomFields{};
+
+    CPL_DISALLOW_COPY_ASSIGN(GDALVectorSelectAlgorithmLayer)
+
+    std::unique_ptr<OGRFeature> TranslateFeature(OGRFeature *poSrcFeature) const
+    {
+        auto poFeature = std::make_unique<OGRFeature>(m_poFeatureDefn);
+        poFeature->SetFID(poSrcFeature->GetFID());
+        const auto styleString = poSrcFeature->GetStyleString();
+        if (styleString)
+            poFeature->SetStyleString(styleString);
+        poFeature->SetFieldsFrom(
+            poSrcFeature, m_anMapSrcFieldsToDstFields.data(), false, false);
+        int iDstGeomField = 0;
+        for (int nSrcGeomField : m_anMapDstGeomFieldsToSrcGeomFields)
+        {
+            poFeature->SetGeomFieldDirectly(
+                iDstGeomField, poSrcFeature->StealGeometry(nSrcGeomField));
+            ++iDstGeomField;
+        }
+        return poFeature;
+    }
+
+    DEFINE_GET_NEXT_FEATURE_THROUGH_RAW(GDALVectorSelectAlgorithmLayer)
+
+  public:
+    explicit GDALVectorSelectAlgorithmLayer(OGRLayer &oSrcLayer)
+        : m_oSrcLayer(oSrcLayer),
+          m_poFeatureDefn(new OGRFeatureDefn(oSrcLayer.GetName()))
+    {
+        SetDescription(oSrcLayer.GetDescription());
+        m_poFeatureDefn->SetGeomType(wkbNone);
+        m_poFeatureDefn->Reference();
+    }
+
+    ~GDALVectorSelectAlgorithmLayer() override
+    {
+        if (m_poFeatureDefn)
+            m_poFeatureDefn->Dereference();
+    }
+
+    bool IncludeFields(const std::vector<std::string> &selectedFields,
```

Comment:
The only thought I have on the implementation is whether the processing could be handled by creating an in-memory VRT that could either be materialized or not, depending on the output format. I don't think anything about this interface precludes that, so I could look into it later on as part of https://github.com/OSGeo/gdal/issues/10066.

Reply:
Same comment could apply to #11827

Reply:
Good question. Regarding VRT'ization of all of this, I believe both GDAL and OGR VRT will have to be extended to support accepting a datasource and a pipeline string, rather than trying to extend the capabilities of the VRT drivers themselves to support all algorithms, which would involve duplicating all the code.

Reply:
ticketed at https://github.com/OSGeo/gdal/issues/11837